### PR TITLE
:sparkles: WIP: new APILifecycle CRD

### PR DESCRIPTION
**Pull request by lionelvillard**
## Summary

Bare minimum implementation of the APILifecycle controller. 

## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+)
+
+// APILifecycle
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apilifecycle.go:25]: this needs much more explanation. It's maybe the most important part of the PR.

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="APIExport reference must not be changed"
+	Reference string `json:"reference"`
+
+	// hooks hold all possible lifecycle hooks
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	Hooks APILifecycleHooks `json:"hooks"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apilifecycle.go:70]: what are these hooks in the future? Am hesitant to want a (core) API to go in without a clear plan where it is heading.

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	// permissionClaims records decisions about permission claims requested by the API service provider.
+	// Individual claims can be accepted or rejected. If accepted, the API service provider gets the
+	// requested access to the specified resources in this workspace. Access is granted per
+	// GroupResource, identity, and other properties.
+	//
+	// +optional
+	PermissionClaims []PermissionClaim `json:"permissionClaims,omitempty"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apilifecycle.go:78]: Where are these coming from? How do they interact with the `APIBinding.spec.claims` ? From the API docs it's hard to understand how both resources interact.

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+// APILifecycleHooks defines the lifecycle hooks
+type APILifecycleHooks struct {
+	// bind is invoked when a binding is created and updated
+	// +optional
+	Bind *APILifecycleHook `json:"bind"`
+}
+
+type APILifecycleHook struct {
+	// url where the hook is located
+	URL string `json:"url"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apilifecycle.go:90]: what is this URL ? a webhook? manifests? What's the payload on post? What does it return?

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	//
+	// +optional
+	PermissionClaims []PermissionClaim `json:"permissionClaims,omitempty"`
+}
+
+// APILifecycleHooks defines the lifecycle hooks
+type APILifecycleHooks struct {
+	// bind is invoked when a binding is created and updated
+	// +optional
+	Bind *APILifecycleHook `json:"bind"`
```
Comment (sttts) [inline pkg/apis/apis/v1alpha1/types_apilifecycle.go:85]: what does "when" mean exactly?

File: pkg/apis/apis/v1alpha1/types_apilifecycle.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// APILifecycleSpec records the APIs and implementations that are to be bound.
+type APILifecycleSpec struct {
+	// reference uniquely identifies an APIExport
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="APIExport reference must not be changed"
+	Reference string `json:"reference"`
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/apis/apis/v1alpha1/types_apilifecycle.go:64]: looking at https://github.com/kcp-dev/kcp/pull/2857/files#diff-8974af955051bac46bc873a4f200e0bb99601d12b165ae055fe0d6f287d12d8dR102 this indicates being the APIExport name; shoudn't this be the export hash instead? I believe we have prior art referencing API exports in kcp that could be reused here.

**Comment by mjudeikis** [comment]
@kcp-dev/kcp-maintainers  do we still want to do this?

**Pull request closed without merging**